File: pricemap-python/pricemap/adapters/mappers/listings.py
from pricemap.adapters.repository.models.listings import ListingModel
from pricemap.domain.entities.listings import ListingEntity


class ListingMapper:
    @staticmethod
    def from_dict_to_entity(data: dict) -> ListingEntity:
        listing = ListingEntity(
            name=data["name"],
            street_address=data["postal_address"]["street_address"],
            postal_code=data["postal_address"]["postal_code"],
            city=data["postal_address"]["city"],
            country=data["postal_address"]["country"],
            description=data["description"],
            building_type=data["building_type"],
            price=data["price"]["price_eur"],
            price_date=data["price"]["date_posted"],
            surface_area_m2=data["surface_area_m2"],
            rooms_count=data["rooms_count"],
            bedrooms_count=data["bedrooms_count"],
        )
        return listing

    @staticmethod
    def from_entity_to_model(listing: ListingEntity) -> ListingModel:
        listing_model = ListingModel(
            name=listing.name,
            street_address=listing.postal_address.street_address,
            postal_code=listing.postal_address.postal_code,
            city=listing.postal_address.city,
            country=listing.postal_address.country,
            description=listing.description,
            building_type=listing.building_type,
            price=listing.price.price_eur,
            surface_area_m2=listing.surface_area_m2,
            rooms_count=listing.rooms_count,
            bedrooms_count=listing.bedrooms_count,
            contact_phone_number=listing.contact_phone_number,
        )
        return listing_model

    @staticmethod
    def from_model_to_dict(listing: ListingModel) -> dict:
        listing_dict = {
            "id": listing.id,
            "name": listing.name,
            "postal_address": {
                "street_address": listing.street_address,
                "postal_code": listing.postal_code,
                "city": listing.city,
                "country": listing.country,
            },
            "description": listing.description,
            "building_type": listing.building_type,
            "price": {
                "price_eur": listing.price,
                "created_date": listing.price_date,
            },
            "surface_area_m2": listing.surface_area_m2,
            "rooms_count": listing.rooms_count,
            "bedrooms_count": listing.bedrooms_count,
            "contact_phone_number": listing.contact_phone_number,
            "created_date": listing.created_date,
            "updated_date": listing.updated_date,
        }
        return listing_dict

File: pricemap-typescript-serverless/src/repositories/listings.ts
import PostgresClient from "serverless-postgres";
import { Listing, ListingReadOnly } from "@/types.generated";
import { extractVariables } from "@/libs/postgres";
import {NotFound} from "@/libs/errors";

type ListingTableRow = {
  id?: number;
  created_date: Date;
  updated_date: Date;
  name: string;
  description: string;
  building_type: Listing["building_type"];
  surface_area_m2: number;
  rooms_count: number;
  bedrooms_count: number;
  contact_phone_number: string;
  price: number;
  street_address: string;
  postal_code: string;
  city: string;
  country: string;
};

function tableRowToListing(row: ListingTableRow): Listing & ListingReadOnly {
  return {
    id: row.id,
    description: row.description,
    name: row.name,
    surface_area_m2: row.surface_area_m2,
    contact_phone_number: row.contact_phone_number,
    building_type: row.building_type,
    rooms_count: row.rooms_count,
    bedrooms_count: row.bedrooms_count,
    latest_price_eur: row.price,
    postal_address: {
      street_address: row.street_address,
      postal_code: row.postal_code,
      city: row.city,
      country: row.country,
    },
    created_date: row.created_date.toISOString(),
    updated_date: row.updated_date.toISOString(),
  };
}

function listingToTableRow(
  listing: Listing,
  createdDate: Date
): ListingTableRow {
  return {
    description: listing.description,
    name: listing.name,
    surface_area_m2: listing.surface_area_m2,
    contact_phone_number: listing.contact_phone_number,
    building_type: listing.building_type,
    rooms_count: listing.rooms_count,
    bedrooms_count: listing.bedrooms_count,
    price: listing.latest_price_eur,
    street_address: listing.postal_address.street_address,
    postal_code: listing.postal_address.postal_code,
    city: listing.postal_address.city,
    country: listing.postal_address.country,
    created_date: createdDate,
    updated_date: new Date(),
  };
}

export function getRepository(postgres: PostgresClient) {
  return {
    async getAllListings(): Promise<(Listing & ListingReadOnly)[]> {
      const queryString = `SELECT * FROM listing`;
      const result = await postgres.query(queryString);

      return result.rows.map(tableRowToListing);
    },

    async getListing(listingId: number) {
      const queryString = `SELECT * FROM listing WHERE id = $1`;
      const queryValues = [listingId];

      const result = await postgres.query(queryString, queryValues);
      const listing = result.rows[0]

      if (!listing) {
        throw new NotFound(`Could not find listing with id : ${listingId}`)
      }

      return tableRowToListing(listing);
    },

    async insertListing(listing: Listing) {
      const tableRow = listingToTableRow(listing, new Date());

      const {
        columns,
        variables,
        values: queryValues,
      } = extractVariables(tableRow);

      const queryString = `
        INSERT INTO listing (${columns.join(",")})
        VALUES(${variables})
        RETURNING *
      `;
      const result = await postgres.query(queryString, queryValues);

      return tableRowToListing(result.rows[0]);
    },

    async updateListing(listingId: number, listing: Listing) {
      const originalListing = await this.getListing(listingId);

      const tableRow = listingToTableRow(listing, originalListing.created_date);
      const { columns, columnsVariables, values } = extractVariables(tableRow);

      const queryString = `
        UPDATE listing
          SET ${columnsVariables.join(", ")}
          WHERE id = $${columns.length + 1}
        RETURNING *
      `;
      const queryValues = [...values, listingId];
      const result = await postgres.query(queryString, queryValues);

      return tableRowToListing(result.rows[0]);
    },
  };
}
